### ISSUE TITLE
Avoid circular dependencies in @grpc/grpc-js

### ISSUE THREAD

**Issue by merlinnot** [grpc/grpc-node#1821]
### Is your feature request related to a problem? Please describe.

While Node.js supports circular dependencies, it comes with some drawbacks and in my experiences it is best to avoid them. As an example, tree shaking with Rollup becomes less effective.

### Describe the solution you'd like

Could we refactor the code in such a way, that there are no import cycles?

### Describe alternatives you've considered

-

### Additional context

-

**Comment by murgatroid99** [comment]
Are you having issues with circular dependencies in the TypeScript code, or in the compiled JavaScript code? The latter is a subset of the former, and looks easier to fix.

**Comment by merlinnot** [comment]
In my case it's JavaScript. Long-term we could look into TypeScript too, I don't know if that's still the case, but TypeScript compiler used to be sluggish when a project had circular dependencies. I'm not sure if that would matter only for contributors to the library or for TypeScript projects having it as a dependency as well.

Anyway, we can take it one step at a time :slightly_smiling_face:

**Comment by tom-andersen** [comment]
Bump. see: https://github.com/firebase/firebase-js-sdk/issues/8231

**Comment by murgatroid99** [comment]
@tom-andersen In that issue, it looks like you are trying to load the library in the browser. That's not going to work anyway. The Node functionality that grpc-js depends on does not exist in the browser.

**Comment by tom-andersen** [comment]
@murgatroid99 You are correct! There appears to be an additional problem of bundling node dependencies on a web build.
